File: ios/AnalyticsModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(AnalyticsModule, AnalyticsModule, NSObject)

RCT_EXTERN_METHOD(initWithConfig:(NSString *)nativeId config:(nullable id)config)
RCT_EXTERN_METHOD(destroy:(NSString *)nativeId)
RCT_EXTERN_METHOD(attach:(NSString *)nativeId playerId:(NSString *)playerId)
RCT_EXTERN_METHOD(detach:(NSString *)nativeId)
RCT_EXTERN_METHOD(setCustomDataOnce:(NSString *)nativeId json:(nullable id)json)
RCT_EXTERN_METHOD(setCustomData:(NSString *)nativeId playerId:(nullable NSString *)playerId json:(nullable id)json)
RCT_EXTERN_METHOD(getCustomData:(NSString *)nativeId playerId:(nullable NSString *)playerId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getUserId:(NSString *)nativeId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/AnalyticsModule.swift
import BitmovinPlayer
import BitmovinCollector

@objc(AnalyticsModule)
class AnalyticsModule: NSObject, RCTBridgeModule {
    /// React bridge reference.
    @objc var bridge: RCTBridge!

    /// PlayerModule instance fetched from the bridge's registry
    @objc var playerModule: PlayerModule? {
        bridge.module(for: PlayerModule.self) as? PlayerModule
    }

    /// JS module name.
    static func moduleName() -> String! {
        "AnalyticsModule"
    }

    /// Module requires main thread initialization.
    static func requiresMainQueueSetup() -> Bool {
        true
    }

    /// Use `UIManager.addBlock` to enqueue module methods on UI thread.
    var methodQueue: DispatchQueue! {
        bridge.uiManager.methodQueue
    }

    /**
     Sends a sample with the provided custom data.
     Does not change the configured custom data of the collector or source.
     - Parameter playerId: Native Id of the player instance.
     - Parameter json: Custom data config json.
     */
    @objc(sendCustomDataEvent:json:)
    func sendCustomDataEvent(_ playerId: NativeId, json: Any?) {
        bridge.uiManager.addUIBlock { [weak self] _, _ in
            guard
                let playerAnalytics = self?.playerModule?.retrieve(playerId)?.analytics,
                let customData = RCTConvert.analyticsCustomData(json)
            else {
                return
            }
            playerAnalytics.sendCustomDataEvent(customData: customData)
        }
    }

    /**
     Gets the current user Id for a `BitmovinPlayerCollector` instance.
     - Parameter playerId: Native Id of the the player instance.
     - Parameter resolver: JS promise resolver.
     - Parameter rejecter: JS promise rejecter.
     */
    @objc(getUserId:resolver:rejecter:)
    func getUserId(
        _ playerId: NativeId,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        bridge.uiManager.addUIBlock { [weak self] _, _ in
            guard let playerAnalytics = self?.playerModule?.retrieve(playerId)?.analytics else {
                reject("[AnalyticsModule]", "Could not find player with ID (\(playerId))", nil)
                return
            }
            resolve(playerAnalytics.userId)
        }
    }
}
